agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2017 Red Hat, Inc. and individual contributors as indicated by the @author tags.
+// You may not use this file except in compliance with the Apache License, Version 2.0.
+
+package io.agroal.api.cache;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * Implementations of ConnectionCache that rely on {@link ThreadLocal}.
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+public interface LocalConnectionCache {
+
+    /**
+     * A local cache that stores at most a single connection.
+     */
+    static ConnectionCache single() {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Acquirable> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Acquirable acquirable = threadLocal.get();
+                return acquirable != null && acquirable.acquire() ? acquirable : null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                if ( acquirable.isAcquirable() ) {
+                    threadLocal.set( acquirable );
+                }
+            }
+
+            @Override
+            public void reset() {
+                threadLocal = new ThreadLocal<>();
+            }
+        };
+    }
+
+    /**
+     * A local cache that stores up to a number of connections.
+     */
+    static ConnectionCache fixed(int size) {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Acquirable[]> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Acquirable[] cacheArray = threadLocal.get();
+                for ( int i = cacheArray.length; i > 0; ) { // iterate from last to first
+                    Acquirable element = cacheArray[--i];
+                    if ( element != null ) {
+                        cacheArray[i] = null;
+                        if ( element.acquire() ) {
+                            return element;
+                        }
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                Acquirable[] cacheArray = threadLocal.get();
```

Comment:
Why does the one above have an isAcquirable() check but not this one?

Reply:
yes, it's missing. Pool calls put on all connection return even when these are to be flushed.